bindings/pyroot/pythonizations/src/GenericPyz.cxx

```diff
@@ -27,13 +29,44 @@ static std::string GetCppName(const CPPInstance *self)
 
 PyObject *ClingPrintValue(CPPInstance *self, PyObject * /* args */)
 {
+   // Map holding the class names and the pointer to the printer
+   // function.
+   static std::unordered_map<std::string, void *> clNamePrinterMap;
+
    auto cppObj = self->GetObject();
    if (!cppObj)
       // Proxied cpp object is null, use cppyy's generic __repr__
       return PyObject_Repr((PyObject*)self);
 
+   // We jit the helper only once, at the first invocation of any
+   // printer
+   if (0 == clNamePrinterMap.size()) {
+      std::string printerCode = "namespace ROOT::Internal::Pythonizations::ValuePrinters"
+                                "{"
+                                "   template<class T> std::string ValuePrinter(void *obj)"
+                                "   {"
+                                "      return cling::printValue((T *)obj);"
+                                "   }"
+                                "}";
+      gInterpreter->Declare(printerCode.c_str());
+   }
+
    const std::string className = GetCppName(self);
-   auto printResult = gInterpreter->ToString(className.c_str(), cppObj);
+
+   // If we never encountered this class, we jit the function which
+   // is necessary to print it and store it in the map instantiated
+   // above. Otherwise, we just use the pointer to the previously
+   // jitted function. This allows to jit the printer only once per
+   // type, at the modest price of a typename and pointer stored in
+   // memory.
+   auto &printerFuncrPtr = clNamePrinterMap[className];
+
+   if (!printerFuncrPtr) {
+      std::string printFuncName = "ROOT::Internal::Pythonizations::ValuePrinters::ValuePrinter<" + className + ">";
+      printerFuncrPtr = (void *)gInterpreter->Calc(printFuncName.c_str());
```

Comment:
```suggestion
      printerFuncrPtr = (void *)gInterpreter->Calc(printFuncName.c_str());
      clNamePrinterMap[className] = printerFuncrPtr;
```
I.e. isn't the insertion currently missing?

Side question: what happens if the user unload code and library between 2 calls to the value printer (in particular if the user unload, change and reload the code for `className`)?

Reply:
First quick answer to the first comment: the assignment is there because at [line 62](https://github.com/root-project/root/pull/13876/files/529a16647b30e700951b91d0b2b100e748cde0b0#diff-02a33032d66f16755503d55293f36a01a92f6e9ab894b5912d61b1d69adbee35R62) a reference to the value is taken.

Reply:
So first of all, we have an issue with unloading in PyROOT, unrelated to this discussion. See these two examples at the prompt:
```cpp
root [0] struct A{vector<int> i{1,2,3};}
root [1] A a1;
root [2] cout << a1.i[0] << endl;
1
root [3] struct A{vector<float> i{4,5,6};}
root [4] A a2;
root [5] cout << a2.i[0] << endl;
4
```
and 
```py
>>> import ROOT
>>> 
>>> ROOT.gInterpreter.ProcessLine("struct A{vector<int> i{1,2,3};}")
0
>>> a1 = ROOT.A()
>>> print(a1.i[0])
1
>>> ROOT.gInterpreter.ProcessLine("struct A{vector<float> i{4,5,6};}")
0
>>> a2 = ROOT.A()
>>> print(a2.i[0])
1
```

Reply:
Alright so we have bigger problems ... but this caching (that will get stale, even plausibly holding tangling pointer in case of unloading) should be looked at if/when we tackle the unloading/reloading issue in cppyy/PyRoot.